lib/puppetserver/ca/action/list.rb

```diff
@@ -79,16 +89,55 @@ def run(input)
           requested, signed, revoked = separate_certs(all_certs)
           missing = certnames - all_certs.map { |cert| cert['name'] }
 
-          (all || certnames.any?) \
-            ? output_certs_by_state(requested, signed, revoked, missing)
-            : output_certs_by_state(requested)
+          (all || certnames.any?)\
+          ? output_certs_by_state(all, output_format, requested, signed, revoked, missing)
+          : output_certs_by_state(all, output_format, requested)
 
           return missing.any? \
             ? 1
             : 0
         end
+        
+        def output_certs_by_state(all, output_format, requested, signed = [], revoked = [], missing = [])
+          if output_format == 'json'
+            output_certs_json_format(all, requested, signed, revoked, missing)
+          else
+            output_certs_text_format(requested, signed, revoked, missing)
+          end
+        end
 
-        def output_certs_by_state(requested, signed = [], revoked = [], missing = [])
+        def output_certs_json_format(all, requested, signed, revoked, missing)
+          if revoked.empty? && signed.empty? && requested.empty? && missing.empty?
+            @logger.inform([].to_json)
+            return
+          end
+
+          if all
+            grouped_cert = { "requested" => requested,
+                             "signed" => signed,
+                             "revoked" => revoked,
+                             "missing" => missing }.to_json
+            @logger.inform(grouped_cert)
+          else
+            unless requested.empty?
+              @logger.inform(requested.to_json)
```

Comment:
So right now the output for various invocations looks like this:
explicit `--certname` flag where the certname does not exist:
```
 ~/Work/server/puppetserver-ca-cli | bundle exec exe/puppetserver-ca list --format text --certname rupert
Missing Certificates:
    rupert
```
vs
```
 ~/Work/server/puppetserver-ca-cli | bundle exec exe/puppetserver-ca list --format json --certname rupert 
["rupert"]
```
`--certname` cert exists and is signed:
```
 ~/Work/server/puppetserver-ca-cli | bundle exec exe/puppetserver-ca list --format text --certname localhost
Signed Certificates:
    localhost       (SHA256)  8A:DE:19:07:93:48:1B:31:B2:BB:FA:37:74:88:D3:DF:53:5A:B2:77:FF:70:97:C8:23:F1:EB:80:9B:53:4F:7C   alt names: ["DNS:puppet", "DNS:localhost"]      authorization extensions: [pp_cli_auth: true]
```
vs
```
 ~/Work/server/puppetserver-ca-cli | bundle exec exe/puppetserver-ca list --format json --certname localhost | jq .
[
  {
    "name": "localhost",
    "dns_alt_names": [
      "DNS:puppet",
      "DNS:localhost"
    ],
    "state": "signed",
    "authorization_extensions": {
      "pp_cli_auth": "true"
    },
    "subject_alt_names": [
      "DNS:puppet",
      "DNS:localhost"
    ],
    "not_before": "2021-06-30T18:05:03UTC",
    "fingerprints": {
      "SHA1": "C1:05:3F:1E:30:B0:C2:A3:FA:95:FB:0F:31:54:38:03:51:2C:FC:8B",
      "SHA256": "8A:DE:19:07:93:48:1B:31:B2:BB:FA:37:74:88:D3:DF:53:5A:B2:77:FF:70:97:C8:23:F1:EB:80:9B:53:4F:7C",
      "SHA512": "06:4F:94:92:B5:3C:8A:AF:58:96:0D:23:8F:E9:85:3B:9E:26:57:0F:1F:12:D1:40:EE:6E:FF:5B:01:D4:FA:57:7E:D4:50:A6:7D:99:F2:79:E0:5D:0F:EE:EF:B0:59:1F:33:16:89:0C:C6:7D:69:EE:05:80:3E:1D:F8:48:42:EE",
      "default": "8A:DE:19:07:93:48:1B:31:B2:BB:FA:37:74:88:D3:DF:53:5A:B2:77:FF:70:97:C8:23:F1:EB:80:9B:53:4F:7C"
    },
    "not_after": "2026-06-30T18:05:03UTC",
    "serial_number": 2,
    "fingerprint": "8A:DE:19:07:93:48:1B:31:B2:BB:FA:37:74:88:D3:DF:53:5A:B2:77:FF:70:97:C8:23:F1:EB:80:9B:53:4F:7C"
  }
]
```
all
```
 ~/Work/server/puppetserver-ca-cli | bundle exec exe/puppetserver-ca list --format text --all
Signed Certificates:
    localhost       (SHA256)  8A:DE:19:07:93:48:1B:31:B2:BB:FA:37:74:88:D3:DF:53:5A:B2:77:FF:70:97:C8:23:F1:EB:80:9B:53:4F:7C   alt names: ["DNS:puppet", "DNS:localhost"]      authorization extensions: [pp_cli_auth: true]
```
vs
```
 ~/Work/server/puppetserver-ca-cli | bundle exec exe/puppetserver-ca list --format json --all | jq .
{
  "requested": [],
  "signed": [
    {
      "name": "localhost",
      "dns_alt_names": [
        "DNS:puppet",
        "DNS:localhost"
      ],
      "state": "signed",
      "authorization_extensions": {
        "pp_cli_auth": "true"
      },
      "subject_alt_names": [
        "DNS:puppet",
        "DNS:localhost"
      ],
      "not_before": "2021-06-30T18:05:03UTC",
      "fingerprints": {
        "SHA1": "C1:05:3F:1E:30:B0:C2:A3:FA:95:FB:0F:31:54:38:03:51:2C:FC:8B",
        "SHA256": "8A:DE:19:07:93:48:1B:31:B2:BB:FA:37:74:88:D3:DF:53:5A:B2:77:FF:70:97:C8:23:F1:EB:80:9B:53:4F:7C",
        "SHA512": "06:4F:94:92:B5:3C:8A:AF:58:96:0D:23:8F:E9:85:3B:9E:26:57:0F:1F:12:D1:40:EE:6E:FF:5B:01:D4:FA:57:7E:D4:50:A6:7D:99:F2:79:E0:5D:0F:EE:EF:B0:59:1F:33:16:89:0C:C6:7D:69:EE:05:80:3E:1D:F8:48:42:EE",
        "default": "8A:DE:19:07:93:48:1B:31:B2:BB:FA:37:74:88:D3:DF:53:5A:B2:77:FF:70:97:C8:23:F1:EB:80:9B:53:4F:7C"
      },
      "not_after": "2026-06-30T18:05:03UTC",
      "serial_number": 2,
      "fingerprint": "8A:DE:19:07:93:48:1B:31:B2:BB:FA:37:74:88:D3:DF:53:5A:B2:77:FF:70:97:C8:23:F1:EB:80:9B:53:4F:7C"
    }
  ],
  "revoked": [],
  "missing": []
}
```
bare invocation (defaults to pending):
```
 ~/Work/server/puppetserver-ca-cli | bundle exec exe/puppetserver-ca list --format json | jq .
[]
```
vs
```
No certificates to list
```
with a pending CSR:
```
 ~/Work/server/puppetserver-ca-cli | bundle exec exe/puppetserver-ca list --format text
Requested Certificates:
    fooberry       (SHA256)  63:70:1E:97:32:BA:CE:AB:70:6B:6B:24:91:92:7B:27:0F:DE:69:F1:64:34:E9:2A:35:E7:24:39:96:2E:56:84
```
vs
```
 ~/Work/server/puppetserver-ca-cli | bundle exec exe/puppetserver-ca list --format json | jq .
[
  {
    "name": "fooberry",
    "state": "requested",
    "dns_alt_names": [],
    "subject_alt_names": [],
    "authorization_extensions": {},
    "fingerprint": "63:70:1E:97:32:BA:CE:AB:70:6B:6B:24:91:92:7B:27:0F:DE:69:F1:64:34:E9:2A:35:E7:24:39:96:2E:56:84",
    "fingerprints": {
      "SHA1": "21:72:43:DE:4A:EB:83:74:02:38:4C:6F:5F:06:E1:D9:DB:AA:04:5D",
      "SHA256": "63:70:1E:97:32:BA:CE:AB:70:6B:6B:24:91:92:7B:27:0F:DE:69:F1:64:34:E9:2A:35:E7:24:39:96:2E:56:84",
      "SHA512": "90:65:D4:11:FD:86:AC:FC:54:EA:5D:67:F7:4B:EA:33:0B:61:1E:04:03:C3:50:B9:4A:68:78:66:11:27:59:04:DC:20:D6:2A:EA:82:D0:6F:F8:AF:FD:D6:4F:2A:23:22:26:C8:2D:A7:F9:91:56:3F:9F:BC:DA:09:C9:3E:25:A5",
      "default": "63:70:1E:97:32:BA:CE:AB:70:6B:6B:24:91:92:7B:27:0F:DE:69:F1:64:34:E9:2A:35:E7:24:39:96:2E:56:84"
    }
  }
]
```

Reply:
So, if you pass `--all` you get a JSON object, but if you don't you get a JSON array. And if you pass `--certname` the contents of the array are either an object representing a cert or a string representing a certname that is missing a cert on the CA. If you don't pass either flag you get a potentially empty list of pending csrs. In both the missing cert and pending csr cases the text output gives you some nice context to say what kind of entity you're dealing with.

Reply:
Is there any trimming or changes I should make?

Reply:
So I think we should have one type of entity returned by all the invocations. And I think it should be a JSON object because they are easier to extend with more information later, also the keys will provide the kind of information that the heading were providing before. eg `{ "missing": [ "rupert" ] }` when `--certname rupert` doesn't exist on the CA.

Reply:
That leaves a little bit of weirdness where we already discussed for `--all` showing keys with empty arrays if there isn't any objects to populate them, eg `{ "revoked": [], "signed": [], "pending": [ {...} ], "missing": [] }`  However, it seems nice that excluding keys is valuable when we're explicitly excluding that class of object from the output, as we do with missing `--certname` certs or the default listing of pending certs.

Reply:
okay I've gotten a little confused.. I'm not really understanding how this method takes `--certname` into account, so I don't see why we wouldn't want to log `grouped_cert` for all cases..

Reply:
All the certs passed in have already been filtered by any `--certname` requirements on line 88 of this PR. But mainly I wanted to clarify about when I would include keys vs exclude keys based on my above suggestion and our previous discussion about having all keys when `--all` was given.

Reply:
I guess I'm saying I would probably include all keys all the time.

Reply:
I'm a little concerned that doing a:
```
puppetserver ca list
{
  "pending": [ <foo csr info>, <bar csr info> ],
  "revoked": [],
  "signed": []
}
```
could be confusing when there _are_ signed and revoked certs, we just don't show them by default.

Reply:
Agreed, this seems bad.

Reply:
hmm okay, now that I look at it, I think that's a good point. I was mostly thinking about a computer parsing this, which would mean it would have to account for nil keys when trying to get at this info, similar to my argument for including all the keys for `all`. whereas the 'text' format is for human consumption.

Based on all that, I'm wondering if it is actually better to keep the original format, where using `--certname` gives you an array of certs and you could just determine the state from that, if you even need it. I definitely understand wanting to return a consistent object type, but I'm wondering if we should be treating the `--all` case vs. the `--certname` case as more different than we have been. I'm not really sure of the use cases for using one or the other.

Reply:
Yeah, I thought that we could make `--certname` return an object and be easier to parse than the sorted-by-state way to show other results by doing
```
puppetserver ca list --certname foo,unknowncert

{
  "foo": <foo cert or csr info>,
  "unknowncert": { "state": "missing" }
}
```
but that's _real_ different in format from the other. I where on the spectrum ease of parsing for specific workflows is worth the user's cognitive overhead of different formats?

Reply:
You make a good point about, users aren't going to see this so the semantic difference between "a key is here because you request that state be output even though it may be empty" and "a key is always going to be here" probably doesn't mean much? I'm not adamant that we need to make that distinction, but I'm inclined to.